File: javascript/packages/linter/src/rules/a11y-disabled-attribute.ts
import { hasAttribute, getTagLocalName } from "@herb-tools/core"
import { DISABLEABLE_ELEMENTS } from "./html-data.js"

import { BaseRuleVisitor } from "./rule-utils.js"
import { ParserRule } from "../types.js"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLOpenTagNode, ParseResult } from "@herb-tools/core"

class DisabledAttributeVisitor extends BaseRuleVisitor {
  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    this.checkDisabledAttribute(node)
    super.visitHTMLOpenTagNode(node)
  }

  private checkDisabledAttribute(node: HTMLOpenTagNode): void {
    if (!hasAttribute(node, "disabled")) return

    const tagName = getTagLocalName(node)
    if (!tagName) return
    if (DISABLEABLE_ELEMENTS.has(tagName)) return

    this.addOffense(
      `The \`disabled\` attribute is only valid on ${[...DISABLEABLE_ELEMENTS].join(", ")}.`,
      node.tag_name!.location,
    )
  }
}

export class A11yDisabledAttributeRule extends ParserRule {
  static ruleName = "a11y-disabled-attribute"
  static introducedIn = this.version("0.10.2")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "warning"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new DisabledAttributeVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/a11y_disabled_attribute.rs
use crate::utils::html_data::DISABLEABLE_ELEMENTS;
use crate::utils::tag_utils::{has_attribute, tag_name_location};

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(DisabledAttributeVisitor);
define_parser_rule!(
  A11yDisabledAttributeRule,
  "a11y-disabled-attribute",
  Warning,
  DisabledAttributeVisitor,
  enabled: false,
  introduced_in: "0.10.2"
);

impl Visitor for DisabledAttributeVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    if has_attribute(node, "disabled") {
      if let Some(tag_name) = node.tag_name.as_ref().map(|token| token.value.to_lowercase()) {
        if !DISABLEABLE_ELEMENTS.contains(&tag_name.as_str()) {
          self.add_offense(
            format!("The `disabled` attribute is only valid on {}.", DISABLEABLE_ELEMENTS.join(", ")),
            tag_name_location(node),
          );
        }
      }
    }

    self.walk_html_open_tag_node(node);
  }
}
